dsp/window/window_test.go

```diff
@@ -8,170 +8,317 @@ import (
 	"fmt"
 	"testing"
 
+	"gonum.org/v1/gonum/cmplxs"
 	"gonum.org/v1/gonum/floats"
-	"gonum.org/v1/gonum/floats/scalar"
 )
 
+const tol = 1e-6
 const (
 	gaussWin = iota
 	tukeyWin
 )
 
-var windowTests = []struct {
-	name    string
-	fn      func([]float64) []float64
-	fnCmplx func([]complex128) []complex128
-	want    []float64
+// these types describe zero, mono, and multi parameter windowing
+// functions for both real and complex data
+//
+// when looking at the truth tables, it is helpful to know that
+// Go treats []complex128{1} as [1+0i].  More specifically, a variable which
+// is of type complex128 to which you assign an untyped constant has a zero
+// imaginary part.
+
+type transformFunc func([]float64) []float64
+type transformFuncComplex func([]complex128) []complex128
+
+type transformer interface {
+	Transform([]float64) []float64
+}
+
+type transformerComplex interface {
+	TransformComplex([]complex128) []complex128
+}
+
+type dualTransformer interface {
+	transformer
+	transformerComplex
+}
```

Comment:
The first two types are only used to define the last one and it's only used once, so I would delete these all and define the dual case where it's used.

Reply:
I agree that the locality of define-at-place-of-use is better.  It's been said that for the gonum projects, often the tests serve as examples.  It is also pragmatic Go to define minimal interfaces.  I think it is better to show them as separate, at the top of the file before the reader's eyes glaze over scrolling past the truth tables.  There is no requirement that a type implement one or the other.  Especially, Values implements one and ValuesComplex the other.

Reply:
The tests serve maybe three purposes, testing the code, showing use of the code and demonstrating idiomatic Gonum usage. Making a single-use interface distant to its use is not idiomatic use.